testy/runtime.py

```diff
@@ -1,25 +1,102 @@
-import threading
+import multiprocessing
 import time
-import studentCode
+import sys
+import traceback
 
-STUDENT_THREAD_NAME = "student_thread"
+import stateManager
+import studentAPI
+
+from runtimeUtil import *
+
+
+# TODO:
+# 0. Set up testing code for the following features.
+# DONE 1. Have student code go through api to modify state.
+# DONE 2. Imposing timeouts on student code (infinite loop, try-catch)
+# 3. Figure out how to kill student thread.
+# 4. Integrate with Bob's socket code: spin up a communication process
+# 5. stateManager throw badThing on processNameNotFound
+# 6. refactor process startup code: higher order function
+# 7. Writeup how all this works
+# 8. Investigate making BadThing extend exception
+
+allProcesses = {}
 
 def runtime():
-  state = [0]
-  runStudentCode(state)
-  while True:
-    time.sleep(1)
-
-def runStudentCode(state):
-  studentThread = threading.Thread(target=runStudentCodeHelper, name=STUDENT_THREAD_NAME, args=(state,))
-  studentThread.daemon = True
-  studentThread.start()
-
-def runStudentCodeHelper(state):
-  nextCall = time.time()
-  while True:
-    studentCode.main(state)
-    nextCall += 1
-    time.sleep(nextCall - time.time())
+  badThingsQueue = multiprocessing.Queue()
+  stateQueue = multiprocessing.Queue()
+  spawnProcess = processFactory(badThingsQueue, stateQueue)
+  restartCount = 0
+  try:
+    spawnProcess(PROCESS_NAMES.STATE_MANAGER, startStateManager)
+    while True:
+      if restartCount >= 5:
+        print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+        print("Too many restarts, terminating")
+        break
+      print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+      print("Starting studentCode attempt: %s" % (restartCount,))
+      spawnProcess(PROCESS_NAMES.STUDENT_CODE, runStudentCode)
+      while True:
+        newBadThing = badThingsQueue.get(block=True)
+        print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+        print(newBadThing)
+        if (newBadThing.event == BAD_EVENTS.STUDENT_CODE_ERROR) or \
```

Comment:
Not major, but the usual Python style I've seen uses

```python
if (condition_1 and
    condition_2):
    pass
```

as opposed to `\` for continuing lines